node/pkg/telemetry/loki.go

```diff
@@ -0,0 +1,128 @@
+
+	gkzap "github.com/go-kit/kit/log/zap"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/client"
+	"github.com/grafana/loki/pkg/logproto"
+	lokiflag "github.com/grafana/loki/pkg/util/flagext"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+)
+
+// ExternalLoggerLoki implements ExternalLogger for the Grafana Loki cloud logging.
+type ExternalLoggerLoki struct {
+	// c is the promtail client.
+	c client.Client
+
+	// labels is the set of labels to be added to each log entry, based on the severity (since severity is one of the labels).
+	labels map[zapcore.Level]model.LabelSet
+
+	// localLogger is the zap localLogger used to log errors generated by the loki adapter. It does not use telemetry.
+	localLogger *zap.Logger
+}
+
+func (logger *ExternalLoggerLoki) log(time time.Time, message json.RawMessage, level zapcore.Level) {
+	lokiLabels := logger.labels[level]
+
+	bytes, err := message.MarshalJSON()
+	if err != nil {
+		logger.localLogger.Error("Failed to marshal log message", zap.Error(err))
+		return
+	}
+	entry := api.Entry{
+		Entry: logproto.Entry{
+			Timestamp: time,
+			Line:      string(bytes),
+		},
+
+		Labels: lokiLabels,
+	}
+
+	logger.c.Chan() <- entry
+}
+
+func (logger *ExternalLoggerLoki) close() error {
+	logger.c.Stop()
+	return nil
+}
+
+// NewLokiCloudLogger creates a new Telemetry logger using Grafana Loki Cloud Logging.
+// skipPrivateLogs: if set to `true`, logs with the field zap.Bool("_privateLogEntry", true) will not be logged by telemetry.
+func NewLokiCloudLogger(ctx context.Context, logger *zap.Logger, url string, productName string, project string, skipPrivateLogs bool, labels map[string]string, opts ...option.ClientOption) (*Telemetry, error) {
+	// The localLogger is used to log errors generated by the loki adapter. It does not use telemetry.
+	localLogger := logger.With(zap.String("component", "loki"))
+
+	// The gkLogger is passed into the loki client, which expects a go-kit logger.
+	gkLogger := gkzap.NewZapSugarLogger(localLogger, zapcore.ErrorLevel)
+
+	reg := prometheus.NewRegistry()
+	m := client.NewMetrics(reg)
+
+	serverURL := flagext.URLValue{}
+	err := serverURL.Set(url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Loki client url: %v", err)
+	}
+
+	cfg := client.Config{
+		URL:                    serverURL,
+		BatchWait:              100 * time.Millisecond,
+		BatchSize:              10,
+		DropRateLimitedBatches: true,
+		Client:                 config.HTTPClientConfig{},
+		BackoffConfig:          backoff.Config{MinBackoff: 1 * time.Millisecond, MaxBackoff: 2 * time.Millisecond, MaxRetries: 3},
+		ExternalLabels:         lokiflag.LabelSet{},
+		Timeout:                1 * time.Second,
+		// TenantID: We are not using the tenantID.
+	}
+
+	clientMaxLineSize := 1024
```

Comment:
Out of curiosity, did you do any checking on line lengths of existing logs? I'm assuming if we log VAAs, it is going to bust this every single time. A sample VAA I have for testing `worm parse` is 2097 characters alone.

Reply:
We are no longer logging VAAs, so 1024 *should* be big enough. 

Loki does peg Prometheus metrics, so we will need to monitor those to see if things are getting dropped.